File: LitebaseSDK/src/java/litebase/SQLColumnListClause.java
// Copyright (C) 2000-2013 SuperWaba Ltda.
// Copyright (C) 2014-2020 TotalCross Global Mobile Platform Ltda.
//
// TotalCross Software Development Kit
//
// This library and virtual machine is distributed in the hope that it will
// be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
//
// SPDX-License-Identifier: LGPL-2.1-only

package litebase;

import totalcross.util.*;

/**
 * Internal use only. Represents a SQL column list clause, like order by or group by. <BR>
 * <i>Note:</i> The select clause has a different class <code>SQLSelectClause</code>, since it has a different complexity.
 */
class SQLColumnListClause
{
   /**
    * The column field list.
    */
   SQLResultSetField[] fieldList = new SQLResultSetField[SQLElement.MAX_NUM_COLUMNS];

   /**
    * The number of fields.
    */
   int fieldsCount;
   
   /**
    * Indicates the index to use when doing a sort operation.
    */
   int index = -1; // juliana@230_29: order by and group by now use indices on simple queries.
   
   /**
    * Indicates that the index to be used is composed or not.
    */
   boolean isComposed; // juliana@230_29: order by and group by now use indices on simple queries.

   // juliana@266_1: corrected a possible AIOOBE when doing a order by or group by in a prepared statement in a table with more than 128 columns 
   // on Java and BlackBerry.
   /**
    * Backup for the tableColIndexes, used in prepared statements.
    */
   short[] fieldTableColIndexesBak; // guich@554_37

   /**
    * Compares two SQL column lists clauses. They can only be considered equal if they list the same column list in the same sequence.
    *
    * @param columnListClause the <code>SQLColumnListClause</code> to compare against.
    * @throws SQLParseException If both column lists do not list the same column sequence.
    */
   public void checkEquality(SQLColumnListClause columnListClause) throws SQLParseException
   {  
      int len = fieldsCount; // the length of the first column list.

      if (len != columnListClause.fieldsCount) // If the length of the column lists are different, the lists are different.
         throw new SQLParseException(LitebaseMessage.getMessage(LitebaseMessage.ERR_ORDER_GROUPBY_MUST_MATCH));

      SQLResultSetField[] fieldList2 = columnListClause.fieldList;

      // If a field of one list has an index of a column different of the same member of the other list or they don't match concerning the ordering 
      // of the result, the lists are considered to be different.
      while (--len >= 0)
         if (fieldList[len].tableColIndex != fieldList2[len].tableColIndex)
            throw new SQLParseException(LitebaseMessage.getMessage(LitebaseMessage.ERR_ORDER_GROUPBY_MUST_MATCH));
   }
   
   /**
    * Checks if the column list contains the given column.
    *
    * @param colIndex The column index of the column being searched for.
    * @throws SQLParseException If the column is not in the column list clause.
    */
   void sqlcolumnlistclauseContains(int colIndex) throws SQLParseException
   {
      SQLResultSetField[] list = fieldList;
      int i = fieldsCount;
      while (--i >= 0)
         if (list[i].tableColIndex == colIndex)
            return;
      throw new SQLParseException(LitebaseMessage.getMessage(LitebaseMessage.ERR_AGGREG_FUNCTION_ISNOT_ON_SELECT));
   }

   /** 
    * Binds the column information of the underlying order or group by clause to the select clause. 
    *
    * @param names2Index The select clause columns hash table.
    * @param columTypes The select clause tables column types.
    * @param tableList The select clause tables.
    * @throws SQLParseException If the column in a group or order by clause is not in the select clause or there is a column of type blob in the 
    * clause.
    */
   void bindColumnsSQLColumnListClause(IntHashtable names2Index, byte[] columnTypes, SQLResultSetTable[] tableList) throws SQLParseException
   {
      int i = fieldsCount,
          index;
      SQLResultSetField field;
      
      if (tableList == null) // Bind during a sorting.
         while (--i >= 0)
         {
            field = fieldList[i];
            field.dataType = columnTypes[field.tableColIndex = names2Index.get(field.tableColHashCode, -1)];
         }
      else // Bind before executing the query.  
      {
         int j,
         n = tableList.length;
         
         while (--i >= 0)
         {
            field = fieldList[i];
            index = -1;
            j = n;
            while (--j >= 0)
            {
               index = tableList[j].table.htName2index.get(field.tableColHashCode, -1);
               if (field.alias.indexOf('.') > 0 && !field.alias.startsWith(tableList[j].aliasTableName))
                  index = -1;
               if (index != -1)
               {  
                  field.table = tableList[j].table; // juliana@230_29
                  break;
               }
            }
            
            if (index == -1)
               throw new SQLParseException(LitebaseMessage.getMessage(LitebaseMessage.ERR_UNKNOWN_COLUMN) + field.tableColName);
   
            field.dataType = columnTypes[field.tableColIndex = index];
            if (field.dataType == SQLElement.BLOB)
               throw new SQLParseException(LitebaseMessage.getMessage(LitebaseMessage.ERR_BLOB_ORDER_GROUP));
         }
         
      }
      
   }
   
   // juliana@230_29: order by and group by now use indices on simple queries.
   /**
    * Finds the best index to use in a sort operation.
    */
   void findSortIndex()
   {
      int length = fieldsCount;
      SQLResultSetField[] fields = fieldList;
      SQLResultSetField field = fields[0];
      Table table = field.table;

      if (length == 1)
      {
         // To use an index for ordering, it must use only non-null columns because Litebase indices don't store nulls.
         // If there is only one field and it is a primary key, uses the primary key index (it is not null).
         if (table.primaryKeyCol == field.tableColIndex)
         {
            index = field.tableColIndex;
            isComposed = false;
         }
         
         // If it is another not null field, try to find an index for it.
         else if ((table.columnAttrs[field.tableColIndex] & Utils.ATTR_COLUMN_IS_NOT_NULL) != 0)
         {
            field.findMaxMinIndex();
            index = field.index;
            isComposed = field.isComposed;
         }
      }
      else
      {
         boolean isAscending = field.isAscending,
                 areAllNotNull = true;
         
         while (--length > 0)
         {
            // All the fields to be sorted must have the same table and ordering.
            if ((field = fields[length]).isAscending != isAscending || field.table != table)
               return;
            
            // To use an index for ordering, it must use only non-null columns because Litebase indices don't store nulls.
            if ((table.columnAttrs[field.tableColIndex] & Utils.ATTR_COLUMN_IS_NOT_NULL) == 0)
               areAllNotNull = false;
         }
         
         int i = -1;
         byte[] composedPKCols = table.composedPrimaryKeyCols;
         
         // Checks if the fields to be sorted are the first part of the composed PK.
         if (table.numberComposedPKCols >= (length = fieldsCount))
         {
            while (++i < length)
               if (composedPKCols[i] != fields[i].tableColIndex)
                  break;
            
            if (i == length) // If so, the composed PK can be used (it is not null).
            {
               index = table.composedPK;
               isComposed = true;
               return;
            }
               
         }
         
         // If the fields to be sorted are not part of a composed PK and are not all not null, does not use an index.
         if (!areAllNotNull)
            return;
         
         // If they are all not null, it is necessary to find a composed index for them.
         ComposedIndex[] compIndices = table.composedIndices;
         ComposedIndex compIndex;
         int j;
         
         i = table.numberComposedIndices;
         while (--i >= 0)
         {
            // Checks if the fields to be sorted are the first part of the composed index.
            if ((compIndex = compIndices[i]).columns.length >= length)
            {
               j = -1;
               while (++j < length)
                  if (compIndex.columns[j] != fields[j].tableColIndex)
                     break;
               
               if (j == length) // If so, the composed PK can be used (it is not null).
               {
                  index = i;
                  isComposed = true;
                  return;
               }
            }
         }
      }
   }
}

File: LitebaseSDK/src/native/parser/SQLColumnListClause.c
/*********************************************************************************
 *  TotalCross Software Development Kit - Litebase                               *
 *  Copyright (C) 2000-2011 SuperWaba Ltda.                                      *
 *  All Rights Reserved                                                          *
 *                                                                               *
 *  This library and virtual machine is distributed in the hope that it will     *
 *  be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of    *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                         *
 *                                                                               *
 *********************************************************************************/



/**
 * Defines functions to deal with a SQL column list clause, like order by or group by.
 */

#include "SQLColumnListClause.h"

/**
 * Compares two SQL column lists clauses. They can only be considered equal if they list the same column list in the same sequence.
 *
 * @param clause1 The first list used in the comparison.
 * @param clause1 The second list used in the comparison.
 * @return <code>true</code>, if both column lists list the same column sequence; <code>false</code>, otherwise.
 */
bool sqlcolumnlistclauseEquals(SQLColumnListClause* clause1, SQLColumnListClause* clause2)
{
   TRACE("sqlcolumnlistclauseEquals")
	int32 length = clause1->fieldsCount; // the length of the first column list.
   SQLResultSetField** fieldList1 = clause1->fieldList;
   SQLResultSetField** fieldList2 = clause2->fieldList;

   if (length != clause2->fieldsCount) // If the length of the column lists are different, the lists are different.
      return false;

   // If a field of one list has an index of a column different of the same member of the other list or they don´t match concerning the ordering 
   // of the result, the lists are considered to be different.
   while (--length >= 0)
      if ((*fieldList1++)->tableColIndex != (*fieldList2++)->tableColIndex)
         return false;
   return true;
}

/**
 * Checks if the column list contains the given column.
 *
 * @param clause The column list clause.
 * @param colIndex The column index of the column being searched for.
 * @return <code>true</code> if the column is in the column list clause; <code>false</code>, otherwise.
 */
bool sqlcolumnlistclauseContains(SQLColumnListClause* clause, int32 colIndex)
{
	TRACE("sqlcolumnlistclauseContains")
   int32 i = clause->fieldsCount;
   SQLResultSetField** fieldList = clause->fieldList;
   
   while (--i >= 0)
      if ((*fieldList++)->tableColIndex == colIndex)
         return true;

   return false;
}

/** 
 * Binds the column information of the underlying order or group by clause to the select clause. 
 *
 * @param context The thread context where the function is being executed.
 * @param clause The column list clause.
 * @param names2Index The select clause columns hash table.
 * @param columTypes The select clause tables column types.
 * @param tableList The select clause tables.
 * @param tableListSize The number of tables of the select clause.
 * @throws SQLParseException If the column in a group or order by clause is not in the select clause or there is a column of type blob in the 
 * clause.
 */
bool bindColumnsSQLColumnListClause(Context context, SQLColumnListClause* clause, Hashtable* names2Index, int16* columnTypes, 
                                                                                  SQLResultSetTable** tableList, int32 tableListSize)
{
	TRACE("bindColumnsSQLColumnListClause")
   int32 i = clause->fieldsCount,  
         index = -1;
   SQLResultSetField** fieldList = clause->fieldList;
   SQLResultSetField* field;

   if (tableList) // Binds before executing the query.  
   {
      SQLResultSetTable* rsTable;
      int32 j;

      while (--i >= 0)
      {
         field = fieldList[i];
         j = tableListSize;

         while (--j >=0)
         {
            index = TC_htGet32Inv(&(rsTable = tableList[j])->table->htName2index, field->tableColHashCode);
            if (xstrchr(field->alias, '.') && xstrncmp(field->alias, rsTable->aliasTableName, xstrlen(rsTable->aliasTableName)))
               index = -1;
            if (index >= 0) 
               break;
         }

         if (index < 0)
         {
            TC_throwExceptionNamed(context, "litebase.SQLParseException", getMessage(ERR_UNKNOWN_COLUMN), field->tableColName);
            return false;
         }

         if ((field->dataType = (int8)columnTypes[field->tableColIndex = index]) == BLOB_TYPE)
         {
            TC_throwExceptionNamed(context, "litebase.SQLParseException", getMessage(ERR_BLOB_ORDER_GROUP));
            return false;
         }
      }
   }
   else // Bind during a sorting.
   {
      while (--i >= 0)
      {
         index = TC_htGet32Inv(names2Index, (field = fieldList[i])->tableColHashCode);
         field->dataType = (int8)columnTypes[field->tableColIndex = index];
      }
   }
   
   return true;
}
